gobblin-metrics/src/main/java/gobblin/metrics/reporter/ScheduledReporter.java

```diff
@@ -93,18 +93,14 @@ public static Config setReportingInterval(Config config, long reportingInterval,
     return config.withValue(REPORTING_INTERVAL, ConfigValueFactory.fromAnyRef(seconds + "S"));
   }
 
-  private final ScheduledExecutorService executor;
-  private final MetricFilter metricFilter;
+  private ScheduledExecutorService executor;
+  private MetricFilter metricFilter;
 
   private Optional<ScheduledFuture> scheduledTask;
   private int reportingPeriodSeconds;
 
   public ScheduledReporter(String name, Config config) {
     super(name, config);
-    this.executor = Executors.newSingleThreadScheduledExecutor(
-        ExecutorsUtils.newDaemonThreadFactory(Optional.of(log), Optional.of("metrics-" + name + "-scheduler")));
-    this.reportingPeriodSeconds = parsePeriodToSeconds(
-        config.hasPath(REPORTING_INTERVAL) ? config.getString(REPORTING_INTERVAL) : DEFAULT_REPORTING_INTERVAL_PERIOD);
     this.metricFilter = createMetricFilter(this.config);
```

Comment:
I see, but won't `metricFilter` hit the same issue that `executor` and `reportingPeriodSeconds` were facing? It won't be initialized in time and will probably hit a NPE.

While not the most elegant solution, we could set `this.metricFilter = createMetricFilter(this.config);` in both places and modifying `createMetricFilter` to include the below code:

```
synchronized (this.metricFilterLock) {
  if (this.metricFilter == null) {
    this.metricFilter = createMetricFilter(this.config);
  }
}
```